12-version_control.Rmd

```diff
@@ -0,0 +1,821 @@
+or revert to earlier versions of the project. 
+Luckily Git can work with shorter versions of the commit hash, 
+which is what we have shown in Figure \@ref(fig:vc1-no-changes)
+
+## Commiting changes to a local repository's history
+
+After working on files in your project's local version control repository 
+(e.g., using Jupyter) and saving your work,
+these changes will only initially exist in the
+working directory of your local computer (Figure \@ref(fig:vc2-changes)).
+
+```{r vc2-changes, fig.cap = 'Conceptual map of the Git and GitHub version control systems.', fig.retina = 2}
+image_read("img/vc2-changes.png") %>%
+  image_crop("3632x2000")
+```
+
+Once you reach a point that you want Git to keep a record 
+of the current version of your work, you need to *commit* 
+(*i.e.*, snapshot) your changes. A pre-requisite to this is telling Git which
+files should be included in that snapshot. We call this step **adding** the 
+files to the staging area. 
+Note that the staging area is **not** a real physical location on your computer, 
+it is instead a conceptual placeholder for these files until they are commit.
+The benefit of the Git version control system using a staging area is that you 
+can choose to commit changes from only certain files. For example, 
+in Figure \@ref(fig:vc-ba2-add) we only add the two files 
+that are quite relevant to the analysis project 
+(`analysis.ipynb` and `README.md`) and not our personal scratch notes for the 
+project (`notes.txt`).
+
+```{r vc-ba2-add, fig.cap = 'The modified files for which you want a snapshot need to be **added** to the staging area.', fig.retina = 2}
+image_read("img/vc-ba2-add.png") %>%
+  image_crop("3632x1200")
+```
+
+Once the files we wish to commit to the repository history have been added 
+to the staging area, we can then commit those flagged files to the repository.
+When we do this, we are required to include a helpful *commit message* to tell 
+collaborators (which included future you) about the changes that were made 
+(Figure \@ref(fig:vc-ba3-commit)). 
+It is important to note here that these changes are only being committed to the 
+repository history on you local computer. 
+The repository on GitHub has not changed, and your collaborators would not yet
+be able to see your work.
+
+```{r vc-ba3-commit, fig.cap = "The files in the staging area need a commit message to be **committed** (*i.e.*, snapshot and archived) to the repository's history on your local computer.", fig.retina = 2}
+image_read("img/vc-ba3-commit.png") %>%
+  image_crop("3632x1100")
+```
+
+### Pushing changes from a remote repository
+
+Finally, once the commits that you want to share with your collaborators 
+have been made, you need to *push* (*i.e.*, send) those commits 
+back to GitHub. This updates the repository history on the remote repository 
+(*i.e.*, GitHub) as well as the version of the files you see when you first 
+visit the repository on GitHub (the remote repository working directory)
```

Comment:
same comment about concrete example

Reply:
Not sure what you mean here @leem44 - can you please elaborate?

Reply:
addressed.